baseband/gsb/tests/test_gsb.py

```diff
@@ -206,58 +207,118 @@ def test_decoding(self):
-                                                 payloadsize=payload.size)
-                assert np.all(phased.data == payload.data[:, np.newaxis])
+    def test_payload(self):
+        with open(SAMPLE_RAWDUMP, 'rb') as fh:
+            # For rawdump, payloadsize = (
+            #   8192 samples/frame * 0.5 bytes/sample)
+            payload1 = gsb.GSBPayload.fromfile(fh, payloadsize=2**12)
+            # Values from reading original raw file.
+            assert np.all(payload1.data[:20].ravel() == np.array(
+                [0., -2., -2., 0., 4., -1., -2., -1., 1., 2., -1., 1.,
+                 -1., 1., -2., 0., -1., -2., 1., -1.], dtype=np.float32))
+            assert payload1.data.shape == (8192, 1)
+            assert payload1.sample_shape == (1,)
+            assert payload1.sample_shape.nchan == 1
+            # Passing ``None`` to `payloadsize`throws an error.
+            with pytest.raises(ValueError):
+                gsb.GSBPayload.fromfile(fh, payloadsize=None)
+            payload2 = gsb.GSBPayload.fromdata(payload1.data, bps=4)
+            assert np.all(payload2.data == payload1.data)
+            payload3 = gsb.GSBPayload(payload1.words, bps=4,
+                                      sample_shape=payload1.sample_shape)
+            assert np.all(payload3.data == payload1.data)
+
+        with open(SAMPLE_PHASED[0][0], 'rb') as fh:
+            # For 1 file in phased, payloadsize = (
+            #   4 full samples/file * 512 channels/full sample *
+            #   2 complex elements/channel * 1 bytes/element)
+            payload4 = gsb.GSBPayload.fromfile(fh, payloadsize=2**12, bps=8,
+                                               complex_data=True)
+            # Values from reading pol0, thread0 phased file as if it were raw
+            # (using dummy timestamp file).
+            assert np.all(payload4.data[:20].ravel() == np.array(
+                [30. + 12.j, -1. + 8.j, 7. + 19.j, -25. - 5.j, 26. + 14.j,
+                 -9. + 0.j, -4. - 1.j, 7. + 6.j, 3. + 5.j, 1. - 2.j,
+                 1. - 5.j, 10. - 6.j, 15. - 11.j, -6. + 13.j, 7. + 0.j,
+                 -10. - 1.j, -8. + 7.j, 13. + 7.j, -1. + 1.j, 0. + 4.j],
+                dtype=np.complex64))
+            assert payload4.data.shape == (2048, 1)
+            assert payload4.sample_shape == (1,)
+            assert payload4.sample_shape.nchan == 1
+            payload5 = gsb.GSBPayload.fromdata(payload4.data, bps=8)
+            assert np.all(payload5.data == payload4.data)
+            payload6 = gsb.GSBPayload(payload4.words, bps=8, complex_data=True,
+                                      sample_shape=payload4.sample_shape)
+            assert np.all(payload6.data == payload4.data)
+
+        # Test encoding then decoding consistency via `fromdata`.
+        payload7 = gsb.payload.GSBPayload.fromdata(payload4.data.real, bps=8)
+        assert np.all(payload7.data == payload4.data.real)
+        # Wrong BPS will flatten high payload values.
```

Comment:
I don't understand how the next test can work!?

Reply:
I think (it's been a while since I wrote this) that this just checks that if an incorrect bps is passed, `fromdata` uses `encode_4bit` rather than `encode_8bit` and you lose any values that can't be expressed in 4 bits.  Since it's more confusing than useful (and the encoders are tested elsewhere) I'll delete it.